---
title: "Strategies for analyzing bisulfite-seq data on CTCF data"
author: "Katarzyna Wreczycka"
output: rmarkdown::github_document
date: "`r Sys.Date()`"
---

```{r, echo = FALSE}
knitr::opts_chunk$set(collapse=TRUE, comment="##", fig.retina=2, fig.path = "diff_meth_figs/diff_meth_CTCF-")
```


# Goal

We compared the performance of different methods for calling differentially methylated cytosines using real bisulfite sequencing experiments where we expect to see changes between samples on certain locations. Stadler and colleagues showed that DNA-binding factors can create low-methylated regions upon binding (Stadler et al., Nature 2011). They further show that the reduced methylation is a general feature of CTCF-occupied sites and if the site is unoccupied, the region on and around the site will have high methylation. This means that if the CTCF occupancy changes between two cell types, we expect to see a change in the methylation levels as well. Armed with this information, we looked for differentially methylated bases on regions that gained or lost CTCF binding between two cell types.

We used CTCF binding sites derived form the ENCODE database for selecting methylation sensitive regions. Since it has beed shown by Maurano et al. Cell Rep. 2015 that DNA methylation changes were only likely to occur with binding site dynamics where the binding site was supported by a CTCF motif, we filtered CTCF peaks tare supported by CTCF DNA motifs from the Factorbook motifs (wgEncodeRegTfbsClusteredV3 ENCODE Mar 2012 Freeze (last updated 2013-07-21)).

# Functions

Load libraries and functions.

```{r}
library("methylKit")
library(tools)
library(readr)
library(genomation)
library(ggplot2)
library(grid)
library(gridExtra)

source("./functions/functions.R")

```

Set global variables for input and output files.

```{r}

# the Supplem. Table 2 from the paper (Stadler et al., Nature 2011)
Table_S2.file = "/home/kwreczy/projects/Strategies_for_analyzing_bisulfite-seq_data/data/Table_S2_Location_of_ChIP-seq_binding_positions_in_19_cell_lines.txt"
# Output directory where the ENCODE RRBS CTCF bed files will be saved
RRBS_CTCF_DIR = "/data/akalin/Base/Encode/Rrbs/wgEncodeHaibMethylRrbs/"
# Output directory where the Factorbook motifs will be saved
Factorbook_DIR = "/data/akalin/kwreczy/projects/Strategies_for_analyzing_bisulfite-seq_data/"
Factorbook.motifs.path = paste0(Factorbook_DIR, "wgEncodeRegTfbsClusteredV3.bed.gz")


```


Download RRBS data for CTCF and CTCF DNA motifs.

```{r}

#' # Download RRBS data for CTCF

download.RRBS.CTCF = function(output.dir, cores=10){
  
  url="http://hgdownload.cse.ucsc.edu/goldenPath/hg19/encodeDCC/wgEncodeHaibMethylRrbs/files.txt"
  download.file(url, 
              destfile=paste0(output.dir, "files.txt"),
              method="wget",
              quiet = FALSE, mode = "w",
              cacheOK = TRUE)

  tbl = read.table(paste0(output.dir, "files.txt"), sep="\t")
  file.names = tbl[,1]
  file.bed = as.character(file.names[which(file_ext(file_path_sans_ext(file.names)) ==  "bed")])
  links = paste0("http://hgdownload.cse.ucsc.edu/goldenPath/hg19/encodeDCC/wgEncodeHaibMethylRrbs/",
               file.bed)
  link2bedfiles = paste0(output.dir, "link2bedfiles.txt")
  write.table(links, file=link2bedfiles,
            quote = FALSE,  sep = "\n", row.names = FALSE, col.names = FALSE)

  command = paste0("cat ", link2bedfiles," | xargs -n 1 -P ", cores,"  curl -O -L")
  print(command)
  system(command)
  
}


# download.RRBS.CTCF(RRBS_CTCF_DIR)


#' # CTCF DNA motifs form the Factorbook database

download.Factorbook = function(output.dir){
  url="http://hgdownload.cse.ucsc.edu/goldenpath/hg19/encodeDCC/wgEncodeRegTfbsClustered/wgEncodeRegTfbsClusteredV3.bed.gz"
  download.file(url, 
              destfile=paste0(output.dir, "wgEncodeRegTfbsClusteredV3.bed.gz"),
              method="wget",
              quiet = FALSE, mode = "w",
              cacheOK = TRUE)
}

# download.Factorbook(Factorbook_DIR)

```

Load filenames of ENCODE RRBS bed files, extract cell lines names the filenames and create a data.frame out of it.

```{r}
remove.nonalpha.upper = function(x){
  require(stringr)
  # Swap out all non-alphanumeric characters.
  x = str_replace_all(x, "[^[:alnum:]]", " ")
  # make all letters big
  upperx = toupper(x)
  # remove white spaces
  gsub(" ", "", upperx, fixed = TRUE)
}
#' Read locations of RRBS files for every cell line
rrbs = read.table("/data/akalin/Base/Encode/Rrbs/wgEncodeHaibMethylRrbs/files.txt", 
                  sep="\t",stringsAsFactors = FALSE, col.names=c("filename", "other"))
rrbs.bed = as.character(rrbs$filename[which(file_ext(file_path_sans_ext(rrbs$filename)) ==  "bed")])
rrbs.bed.uw = rrbs.bed[which(grepl("UwSites", rrbs.bed))]
my.list = strsplit(substring(rrbs.bed.uw, 23),split = "UwSites")
rrbs.cell.lines = unlist(lapply(my.list, function(x) x[[1]]))
rrbs.tbl =data.frame(# filenames
                     bed=rrbs.bed.uw, 
                     # names of cell lines wihout non-alphanumeric characters, white spaces, and capital letters 
                     celline = remove.nonalpha.upper(rrbs.cell.lines),
                     # original names of cell lines (form the files.txt file)
                     celline.o=rrbs.cell.lines,
                     stringsAsFactors = FALSE)

print(head(rrbs.tbl))

```


Load CTCF peaks from the Supplem. Table 2 from the paper (Stadler et al., Nature 2011).

```{r}
#' Read supplementary table from the paper
#' And match cell line names from the paper with cell line names from the rrbs bed files.
supp.tbl2.o = read_delim(Table_S2.file, delim="\t")
# Unify names of cell lines
print(colnames(supp.tbl2.o))
supp.tbl2 = supp.tbl2.o[ , c(1:6, grep('pk', colnames(supp.tbl2.o))) ]

supp.cellines = colnames(supp.tbl2)[7:ncol(supp.tbl2)]
supp.cellines = sapply(strsplit(supp.cellines, "_"), function(x){
            if(length(x)==3){
              paste0(x[1],"_", x[2])
            }else{
              x[1]
            }
  })
supp.cellines = data.frame(cell.line.o=supp.cellines, # original cell line names
                           cell.lines=remove.nonalpha.upper(supp.cellines),
                           stringsAsFactors = FALSE)
print(head(supp.cellines))
supp.tbl2.1 = supp.tbl2
colnames(supp.tbl2.1) = c( colnames(supp.tbl2)[1:6],  supp.cellines$cell.lines)
print(colnames(supp.tbl2.1))

#' Convert peaks form the supplementary table to A GRanges object
supp.tbl2.gr = makeGRangesFromDataFrame(supp.tbl2.1,
                                   keep.extra.columns=TRUE)

print(head(supp.tbl2.gr))

```

Filter CTCF peaks based on overlap with Factorbook CTCF DNA motifs. 

```{r}
Factorbook.motifs = readGeneric(file=Factorbook.motifs.path, 
                                chr = 1, start = 2, end = 3, strand = NULL,
                                sep = "\t", meta.cols = list(TFname=4))
motifs.CTCF = Factorbook.motifs[which(Factorbook.motifs$TFname == "CTCF")]
supp.tbl2.gr.CTCFmotif = subsetByOverlaps(supp.tbl2.gr, motifs.CTCF)

# Almost all CTCF peaks from the supplementary contain a CTCF motif
paste0("Percent of CTCF peaks from Stadler et al that overlap CTCF DNA motifs = ", round(length(supp.tbl2.gr.CTCFmotif) / length(supp.tbl2.gr), 4) * 100, "%")
```

Match cell line names from the Supplem. Table 2 with cell line names from the ENCODE RRBS bed files.

```{r}

# Filter cell lines that are in all sets: in suppl table and rrbs set
# INFO: I get only 9 cell lines
cl.in.all = supp.cellines$cell.lines[
  which(supp.cellines$cell.lines %in% unique(rrbs.tbl$celline))
  ]
print(cl.in.all)

#' Create a data frame with all combinations of cell lines (non-symetric)
#' in rows. It has two columns of names of cell lines.
all.comb <- expand.grid(cellline1=cl.in.all, 
                   cellline2=cl.in.all,
                   stringsAsFactors=FALSE,
                   KEEP.OUT.ATTRS = FALSE)
# Deleting reversed duplicates
comb = all.comb[!duplicated(apply(all.comb,1,function(x) paste(sort(x),collapse=''))),]
# Remove rows where cell line are the same
wh.re = which(comb[,1]==comb[,2])

# Combination of cellline pairs
comb.of.cellline.pairs = comb[-wh.re,]

print("Combination of pairs of cell lines")
print(comb.of.cellline.pairs)
```

Load ENCODE RRBS bed files and convert them into a methylBase objects that contain columns indicating coverage and number of methylated
and unmetylated Cs for every replicate of every cell line.

```{r}

#' Read RRBS bed files into a methylBase object
#'
#' Combine information from RRBS data sets for 2 cell lines (each cell line has 2 replicates)
#' into a methylBase object fom the methylKit library. Coverage and number of methylated
#' and unmetylated Cs from every replicate for each cell line is represented as a separate column.
#'
#' @c1, @c2: characters indicating cell lines
#' @rrbs.tbl: a data frame with columns: name of a bed file (bed) and name of celline (celline)
#' 
#' @return a methylBase object that contains information about coverage, 
#'         number of methylated and unmethylated Cs for every position from given 2 cell lines.
bed2methylBase = function(c1,c2,
                          rrbs.tbl,
                          RRBS_PATH="/data/akalin/Base/Encode/Rrbs/wgEncodeHaibMethylRrbs/"){
  require(methylKit)
  # Get path to bed files
  rrbs.c1.bed = paste0(RRBS_PATH, rrbs.tbl[which(rrbs.tbl$celline == c1),]$bed)
  rrbs.c2.bed = paste0(RRBS_PATH, rrbs.tbl[which(rrbs.tbl$celline == c2),]$bed)
  
  # Read bed files
  methylRawList.obj = methRead(as.list(c(rrbs.c1.bed, rrbs.c2.bed)),
                               sample.id=as.list(c(c1, c1, c2, c2)),
                               assembly="hg19", skip=1,
                               treatment = c(0,0,1,1),
                               pipeline=list(chr.col=1, start.col=2, end.col=3, 
                                             coverage.col=10, freqC.col=11, strand.col=6,
                                             fraction=FALSE))
  # Filtering samples based on read coverage
  filtered.methylRawList=filterByCoverage(methylRawList.obj,lo.count=10,lo.perc=NULL,
                                    hi.count=NULL,hi.perc=99.9)
  # Merging samples: take the bases covered in all samples/replicates
  methylBase.rrbs = unite(filtered.methylRawList, destrand=FALSE)
  return(methylBase.rrbs)
}

#' Based on the data.frame comb.of.cellline.pairs that rows indicate combinations of
#' pairs of cell lines, load RRBS bed files for each combination and combine them
#' in a one object (a methylbase object from the methylKit library)
#
# methylBase.obj.list = mclapply(1:nrow(comb.of.cellline.pairs), function(i){
#   c1=comb.of.cellline.pairs[i,]$cellline1
#   c2=comb.of.cellline.pairs[i,]$cellline2
#   methylBase.obj = bed2methylBase(c1, c2, rrbs.tbl)
#   methylBase.obj
# }, mc.cores=20)
# saveRDS(methylBase.obj.list,"/data/akalin/kwreczy/projects/Strategies_for_analyzing_bisulfite-seq_data/methylBase.obj.list.rds")
methylBase.obj.list = readRDS("/data/akalin/kwreczy/projects/Strategies_for_analyzing_bisulfite-seq_data/methylBase.obj.list.rds")

#' Remove chromosomes that have only few CpG sites, 
#' because otherwise BSmooth has errors
#'
#' @methylBase.obj a methylBase object
#'
#' @return a methylBase object
rm.small.chr.methylBase = function(methylBase.obj){
#     # https://support.bioconductor.org/p/65352/
#     #1. If you're comparing
#     #two groups and there is only a single sample in each group, you're out of
#     #luck; you cannot use the t-stat approach in BSmooth
#     #2.The second error, which has been reported many times, typically happens
#     #when you include a very small chromosome (small = few CpGs), like chrMT.
    len.chr = table(methylBase.obj$chr)
    chr.rm = names(len.chr[which(len.chr<200)])
    bin.rm = methylBase.obj$chr %in% chr.rm
    if(sum(bin.rm)!=0){
          methylBase.obj = methylBase.obj[-which(bin.rm),]
    }
    methylBase.obj$chr = droplevels(methylBase.obj$chr, exclude = 0)
    return(methylBase.obj)
}

methylBase.obj.list = mclapply(methylBase.obj.list, rm.small.chr.methylBase , mc.cores=50)
names(methylBase.obj.list) = paste0(comb.of.cellline.pairs[,1], "_",comb.of.cellline.pairs[,2])

```


Call differentially methylated cytosines using BSmooth, DSS, limma and methylKit. 

```{r}
#' Calculate predicted DM

# Run DSS

# models.diff.dss = lapply(methylBase.obj.list, function(x){ run.DSS(x, difference=5, cores=50) })
# saveRDS(models.diff.dss[[1]], "/data/akalin/kwreczy/projects/Strategies_for_analyzing_bisulfite-seq_data/ctcf.calculateDiffMethDSS1.rds")
# saveRDS(models.diff.dss[[2]], "/data/akalin/kwreczy/projects/Strategies_for_analyzing_bisulfite-seq_data/ctcf.models.dss.rds")

# Run methylKit

#ctcf.models.methylkit = lapply(methylBase.obj.list, function(x) {run.methylkit(methylBase.obj, cores=60, difference=5)})
#models.methylkit.diff = readRDS("/data/akalin/kwreczy/projects/Strategies_for_analyzing_bisulfite-seq_data/ctcf.models.methylkit.all.rds")

# Run BSmooth
# source("./functions/functions.R")
# source("./functions/Bsmooth.R")
# source("./functions/my_dmrFinder.R")
#
# ctcf.bsmooth.smooth.same = lapply(methylBase.obj.list, function(methylBase.obj) {
#                     runBSmooth.smooth(methylBase.obj,
#                              estimate.var="same",
#                              cores=60)
# })
# ctcf.bsmooth.same = mclapply(1:length(ctcf.bsmooth.smooth.same),
#                              function(i){
#                                 print(i)
#                                x = ctcf.bsmooth.smooth.same[[i]]
#                                dmrs= smoothed.dmrFinder(x)
#                                dmrs.l = lapply(dmrs, function(dmr){
#                                  makeGRangesFromDataFrame(dmr,
#                                            keep.extra.columns=TRUE)
#                                })
#                                names(dmrs.l) = names(dmrs)
#                                dmrs.l
#                              }, mc.cores=60)
#saveRDS(ctcf.bsmooth.same,"/data/akalin/kwreczy/projects/Strategies_for_analyzing_bisulfite-seq_data/ctcf.bsmooth.same.rds")
#saveRDS(ctcf.bsmooth.same,"/data/akalin/kwreczy/projects/Strategies_for_analyzing_bisulfite-seq_data/ctcf.models.bsmooth.BSmooth.tstat.same.rds")
# ctcf.bsmooth.smooth.same = readRDS("/data/akalin/kwreczy/projects/Strategies_for_analyzing_bisulfite-seq_data/ctcf.models.bsmooth.BSmooth.tstat.same.rds")
# ctcf.bsmooth.same.defualtqvalue = mclapply(1:length(ctcf.bsmooth.smooth.same),
#                              function(i){
#                                 print(i)
#                                x = ctcf.bsmooth.smooth.same[[i]]
#                                dmrs= smoothed.dmrFinder(x)
#                                dmrs.l = lapply(dmrs, function(dmr){
#                                  makeGRangesFromDataFrame(dmr,
#                                            keep.extra.columns=TRUE)
#                                })
#                                names(dmrs.l) = names(dmrs)
#                                dmrs.l
#                              }, mc.cores=36)
# #saveRDS(ctcf.bsmooth.same.defualtqvalue,"/data/akalin/kwreczy/projects/Strategies_for_analyzing_bisulfite-seq_data/ctcf.models.bsmooth.default.rds")

#https://bioconductor.org/packages/release/bioc/vignettes/limma/inst/doc/usersguide.pdf
#The options trend=TRUE and robust=TRUE are also often helpful when running eBayes, increasing
#power for certain types of data.
# source("./functions/limma.R")
# models.limma.diff  = mclapply(1:length(methylBase.obj.list), 
#                                   function(i) {
#                                       methylBase.obj = methylBase.obj.list[[i]]
#                                       lime = limma.meth(methylBase.obj, transform=TRUE, trend=FALSE, robust=FALSE)
#                                       as(getMethylDiff(lime, difference=5, qvalue = 0.01),"GRanges")
#                                       }, 
#                               mc.cores=length(methylBase.obj.list))
# saveRDS(models.limma.diff, "/data/akalin/kwreczy/projects/Strategies_for_analyzing_bisulfite-seq_data/ctcf.models.limma.26072017.trendFALSErobustFALSE.rds")

```


Load output of tools for calculation of differentially methylates cytosines.

```{r}

# Here are saved differentially methylated regions calculated by the tools as RDS files
models.methylkit.diff = readRDS("/data/akalin/kwreczy/projects/Strategies_for_analyzing_bisulfite-seq_data/ctcf.models.methylkit.all.rds")
models.bsmooth.diff = readRDS("/data/akalin/kwreczy/projects/Strategies_for_analyzing_bisulfite-seq_data/ctcf.models.bsmooth.rds")
models.dss.diff = readRDS("/data/akalin/kwreczy/projects/Strategies_for_analyzing_bisulfite-seq_data/ctcf.models.dss.rds")
models.limma.diff = readRDS("/data/akalin/kwreczy/projects/Strategies_for_analyzing_bisulfite-seq_data/ctcf.models.limma.26072017.trendFALSErobustFALSE.rds")
models.bsmooth.diff =readRDS("/data/akalin/kwreczy/projects/Strategies_for_analyzing_bisulfite-seq_data/ctcf.models.bsmooth.default.rds")


models.perclcomb = mapply(c, models.methylkit.diff, models.bsmooth.diff, SIMPLIFY=FALSE)
models.perclcomb = mapply(c, models.perclcomb, models.dss.diff, SIMPLIFY=FALSE)
models.perclcomb = mapply(c, models.perclcomb, limma=models.limma.diff, SIMPLIFY=FALSE)

#' Revert list structure:
#' Now, first element of the list is a tool and then diff. meth. sites for every cell line separately
fun <-  function(ll) {
    nms <- unique(unlist(lapply(ll, function(X) names(X))))
    ll <- lapply(ll, function(X) setNames(X[nms], nms))
    ll <- apply(do.call(rbind, ll), 2, as.list)
    lapply(ll, function(X) X[!sapply(X, is.null)])
}
models = fun(models.perclcomb)

print(names(models))
print(paste0("model=",names(models)[1], ", pair_cell_lines=(", comb.of.cellline.pairs[1,][1], ", ",comb.of.cellline.pairs[1,][2], ") :"))
print(models[[1]][[1]])

```


We performed pairwise comparisons for each pair in all possible combinations of these 19 cell lines. We defined true positives as the number of CTCF peaks gained/lost between two cell lines which overlap at least one DMC. True negatives are defined as the number of CTCF peaks that do not change between cell lines and do not overlap any DMC although they are covered by RRBS reads. Accordingly, false positives are defined as the number of CTCF peaks that are present in both cell lines but overlap with at least one DMC, while false negatives are defined as peaks that are gained or lost between cell lines but have no DMC.

Calculate TP, TN, FP, FN rates and specificity, accuracy and sensitivity:

* TP: CTCF peak change + DM
* TN: CTCF no peak change + no DM
* FP: CTCF no peak change + DM
* FN: CTCF peak change + no DM
* Spec, Acc, Sens
DM = two given cell lines are have at least 1 differentially methylated cytosines between two given cell lines


We down-sampled the CTCF peaks that do not change to match the number of peaks that change, to have a balanced classification performance, otherwise true negatives overwhelm performance metrics since there are many CTCF peaks that do not change.


```{r}
source("./functions/CTCFfunctions.R")

#' Subset set of peaks with no change
model_names = c(
"methylKit.F.qvalue.MN",
"methylKit.Chisq.qvalue.none",
"methylKit.F.qvalue.none",
"methylKit.Chisq.qvalue.MN",
"dss.qvalue",
"limma",
"bsmooth.same.default"
)

set.seed(112)

rates.cellline.model.sample = 
      mclapply(1:length(model_names), 
              function(model_indx){
                      per_cellline = lapply(1:nrow(comb.of.cellline.pairs), 
                             function(pair_celllines_indx){
                                          accuracy.atleast1DM(methylBase.obj.list[[pair_celllines_indx]], 
                                                              comb.of.cellline.pairs[pair_celllines_indx,], 
                                                              supp.tbl2.gr.CTCFmotif, 
                                                              models[[model_names[model_indx]]][[pair_celllines_indx]],
                                                              sample_no_change_peaks=TRUE)
                       })
                       do.call("rbind", per_cellline)
}, mc.cores=length(model_names))

convert2df.sample = lapply(1:length(rates.cellline.model.sample), function(i){
  a=cbind(rates.cellline.model.sample[[i]], model=model_names[i])
  a = as.data.frame(a)
  indx <- 1:8
  a[indx] <- lapply(a[indx], function(x) as.numeric(as.character(x)))
  a
})
CTCF.stats.subset = do.call("rbind", convert2df.sample)
CTCF.stats.subset$precision = as.numeric(as.character(CTCF.stats.subset$precision))
CTCF.stats.subset$NPV = as.numeric(as.character(CTCF.stats.subset$NPV))
CTCF.stats.subset$recall = as.numeric(as.character(CTCF.stats.subset$recall))

print("statistics for limma")
print(CTCF.stats.subset[which(CTCF.stats.subset$model=="limma"),])

```

Visualization of performance measurements of tools for DMCs detection based on the association between CTCF occupancy with methylation status in cell-type specific manner using the Wang et al data and the RRBS ENCODE data.

```{r, message=FALSE, fig.height=10, fig.width=10}
levels(CTCF.stats.subset$model)[levels(CTCF.stats.subset$model)=="methylKit.F.qvalue.MN"] <- "methylKit-Ftest-OC"
levels(CTCF.stats.subset$model)[levels(CTCF.stats.subset$model)=="methylKit.Chisq.qvalue.MN"] <- "methylKit-Chisqtest-OC"
levels(CTCF.stats.subset$model)[levels(CTCF.stats.subset$model)=="methylKit.Chisq.qvalue.none"] <- "methylKit-Chisqtest"
levels(CTCF.stats.subset$model)[levels(CTCF.stats.subset$model)=="limma.qvalue"] <- "limma"
levels(CTCF.stats.subset$model)[levels(CTCF.stats.subset$model)=="dss.qvalue"] <- "DSS"
levels(CTCF.stats.subset$model)[levels(CTCF.stats.subset$model)=="bsmooth.same.default"] <- "BSmooth"

CTCF.stats.subset$model=relevel(CTCF.stats.subset$model, "methylKit-Ftest-OC")
CTCF.stats.subset$model=relevel(CTCF.stats.subset$model, "methylKit-Chisqtest-OC")
CTCF.stats.subset$model=relevel(CTCF.stats.subset$model, "methylKit-Chisqtest")
CTCF.stats.subset$model=relevel(CTCF.stats.subset$model, "limma")
CTCF.stats.subset$model=relevel(CTCF.stats.subset$model, "DSS")
CTCF.stats.subset$model=relevel(CTCF.stats.subset$model, "BSmooth")

# 1. Create the plots
#++++++++++++++++++++++++++++++++++

# The palette 
cbPalette <- c("#999999", "#E69F00", "#56B4E9", "#009E73", "#F0E442", "#0072B2", "#D55E00", "#CC79A7")

p_sens<-ggplot(CTCF.stats.subset, aes(x=model, y=sens, fill=model)) +
  geom_boxplot()+
  coord_cartesian(ylim=c(0.0,1.00))+
  labs(y="Sensitivity\n", x="Tool",fill='Tool')+
  scale_fill_manual(values=cbPalette)+
  theme(axis.text.x  = element_text(angle=45, vjust=1, hjust=1))

p_spec<-ggplot(CTCF.stats.subset, aes(x=model, y=spec, fill=model)) +
  geom_boxplot()+
  coord_cartesian(ylim=c(0.0,1.00))+
  labs(y="Specificity", x="Tool",fill='Tool')+
  scale_fill_manual(values=cbPalette)+
  theme(legend.position = "none", axis.text.x  = element_text(angle=45, vjust=1, hjust=1))

p_fscore<-ggplot(CTCF.stats.subset, aes(x=model, y=f_score, fill=model)) +
  geom_boxplot()+
  coord_cartesian(ylim=c(0.0,1.00))+
  labs(y="F-score", x="Tool",fill='Tool')+
  scale_fill_manual(values=cbPalette)+
  theme(legend.position = "none", axis.text.x  = element_text(angle=45, vjust=1, hjust=1))

# 2. Save the legend
#+++++++++++++++++++++++
get_legend<-function(myggplot){
  tmp <- ggplot_gtable(ggplot_build(myggplot))
  leg <- which(sapply(tmp$grobs, function(x) x$name) == "guide-box")
  legend <- tmp$grobs[[leg]]
  return(legend)
}
legend <- get_legend(p_sens)
# 3. Remove the legend from the box plot
#+++++++++++++++++++++++
p_sens <- p_sens+ theme(legend.position="none")

p_sens1 <- arrangeGrob(p_sens, top = textGrob("a", x=unit(0, "npc"),y=unit(1, "npc"),just=c("left","top"), gp=gpar(col="black", fontsize=14)))
p_spec1 <- arrangeGrob(p_spec, top = textGrob("b", x=unit(0, "npc"),y=unit(1, "npc"),just=c("left","top"), gp=gpar(col="black", fontsize=14)))
p_fscore1 <- arrangeGrob(p_fscore, top = textGrob("c", x=unit(0, "npc"),y=unit(1, "npc"),just=c("left","top"), gp=gpar(col="black", fontsize=14)))

grid.arrange(p_sens1, p_spec1, 
             p_fscore1, legend,
             ncol = 2, nrow=2)
```



```{r}
# Rename names of tools
old.v=c("dss.qvalue", "limma","methylKit.Chisq.qvalue.MN","methylKit.Chisq.qvalue.none", "methylKit.F.qvalue.MN", "methylKit.F.qvalue.none","bsmooth.same.default")
new.v=c("DSS",      "limma"   ,"methylKit-Chisqtest-OC", "methylKit-Chisqtest", "methylKit-Ftest-OC", "methylKit-Ftest","BSmooth")
names(new.v) = old.v

models.perclcomb.selected = lapply(1:length(models.perclcomb),
                                   function(i){
                              x=models.perclcomb[[i]] [ which(names(models.perclcomb[[i]]) %in% model_names) ]
                              names( x ) = as.character(new.v[match(names(x), old.v)])
                              x
                             })

source("./functions/vis.R")
rownames(comb.of.cellline.pairs) = paste0(comb.of.cellline.pairs[,1], "_", comb.of.cellline.pairs[,2])

for(i in 1:length(models.perclcomb.selected)){
  print(rownames(comb.of.cellline.pairs)[i] )
  plot.intersetion.tools( models.perclcomb.selected [[1]] )
}

```


```{r}
sessionInfo()
```


